src/poset/poset.go

```diff
@@ -169,8 +165,8 @@ func (p *Poset) processKnownRoots() *election.ElectionRes {
 // Event order matter: parents first.
 // ProcessEvent is not safe for concurrent use.
 func (p *Poset) ProcessEvent(e *inter.Event) error {
-	if e.Epoch != p.EpochN {
-		return epoch_check.ErrNotRecent
+	if err := epoch_check.New(&p.dag, p).Validate(e); err != nil {
```

Comment:
consider to cache an epochCheck (epoch_check.New(&p.dag, p)) per poset.
Line 58, 168: epoch_check.New(&p.dag, p).Validate(e) => possibly store a local boolean in the poset for e, if validation takes time.